lib/ecto/query.ex

```diff
@@ -980,6 +992,69 @@ defmodule Ecto.Query do
     [on: expr]
   end
 
+  @doc """
+  A recursive common table expression (CTE) also known as WITH expression.
+
+  For keyword queries use `:with` or `:with_recursive` keywords.
```

Comment:
I would rather have `cte` and `recursive_cte`. With two functions, in order to keep everything consistent. :)